### ISSUE TITLE
doRollover() equivalent

### ISSUE THREAD

**Issue by NaruZosa** [Delgan/loguru#699]
Hi there, love Loguru, it makes everything so much easier!

With the existing logging we have set up for a particular tool, it uses doRollover() on the file handler if certain conditions are met (these conditions don't relate to the log file itself, but rather to internal logic).
The main time this occurs is when this tool has a new run start, as each log file then corresponds with a unique run. 

We already have the logic in there, what would we need to do for the equivalent of a doRollover() so we can use Loguru as a drop in replacement for the built in logger?

**Comment by Delgan** [comment]
Hi @BradleyDS2. :) 

Can you use [a custom `rotation` function](https://loguru.readthedocs.io/en/stable/api/logger.html#file) maybe?

```python
class DoRollover:
    def __init__(self):
        self._should_rotate = False

    def rotate(self, message, file):
        ...  # Check internal conditions.
        return False  # Or 'True'.

rotator = DoRollover()
logger.add("file.log", rotation=rotator.rotate)
```

**Comment by BradleyDS2** [comment]
Thank you, that works. Sadly it requires moving all the existing checks into the custom rotation class, increasing the complexity.

Is there a way to call the rotation from another function, such as logger.doRollover() or is this a planned feature?

My apologies if I've overlooked something in the documentation.

**Comment by Delgan** [comment]
My bad, I overlooked what `doRollover()` is actually used for, sorry.

So, rotation possibly occurs when a message is logged, it can't be triggered manually and I'm not sure I want to change that.

The best workaround I can think for is something as follow:

```python
class Rotator:
    def __init__(self):
        self._should_rotate = False

    def do_rollover(self):
        self._should_rotate = True

    def rotate(self, message, file):
        if self.should_rotate:
            self.should_rotate = False
            return True
        return False

rotator = Rotator()
logger.add("file.log", rotation=rotator.rotate)
logger.info("No rotation")
rotator.do_rollover()  # Possibly in another module.
logger.info("Rotation occurs now")
 ```

**Comment by BradleyDS2** [comment]
I can understand not wanting to change rotation to occur before a message is logged, and that's easy enough for me to work around by immediately logging a message stating "Log file rollover occuring" or something similar.

Adding that as a function to the class seems to work, but (this is my personal opinion) feels a bit clunky, as the call to rotate now is being done to the rotator rather than the log file handler.

It also means if someone wants to use do_rollover they need to specify a custom rotator, instead of relying on the fantastic built-in plain English parser such as "10 MB".

In the end it ends up adding a lot of boilerplate, which is something Loguru is generally really good at reducing instead of increasing.


Is this feature something you'd be willing to look at natively implementing? The rollover wouldn't need to occur immediately, but it'd be great to be able to just specify file_logger.rotate_now() without requiring the extra boilerplate, especially as it's available via the native logger library.


Fantastic work you're doing regardless, even with the boilerplate to add the rotate_now, Loguru is so much easier to use than the native logger! I'm changing my tools to use it, because the built in one sometimes requires hacky and unreliable solutions to issues (such as disconnecting the Flask logger) but I haven't experienced any such issues with Loguru

**Comment by Delgan** [comment]
Thanks for your kind words! I'm gland you like Loguru! :smile: 

I agree the proposed workaround isn't ideal compared to a `do_rollover()` method. However, implementing the solution you suggest is not easy either because there can be several file handlers configured, and it's not clear which on `logger.do_rollover()` should rotate. That implies somehow giving access to the internal handlers while they are deliberately private currently.

Another workaround comes to mind: would that be possible for you to call `logger.remove(id)` and `id = logger.add("file.log", rotation="10 MB, ...)`? It should have the same effect as a rotation, actually.

**Comment by BradleyDS2** [comment]
Thank you for the suggestion, I will give it a go next week and report back.

**Comment by BradleyDS2** [comment]
Sorry, haven't had the opportunity to check this yet. Will report in once I do.

**Comment by NaruZosa** [comment]
I sincerely apologise for the delay in report8nt back, things on my end have been hectic.


I've used your solution above, with a tweak to only rotate if the current log file is >0 bytes. It works perfectly for my needs.

I am incredibly grateful for your assistance and fantastic tool, thank you so much ❤️

**Comment by Delgan** [comment]
Haha, no problem, glad it worked. :+1:
